### ISSUE TITLE
Support installer-paths

### ISSUE THREAD

**Issue by xendk** [stevenremot/ede-php-autoload#15]
Drupal 8 (and I suspect other frameworks) uses https://github.com/composer/installers to relocate where the installed packages is placed. 

The bad news is that composer basically allows composer plugins to mess with where packages are installed. The good news is that a large amount of frameworks uses composer/installers rather than their own custom one.

Sadly, composer doesn't note the installed path of a package in the lock file. That would have been too easy.

As I see it, what is needed is:

1. Triggering on composer/installers being part of the dependencies.
2. For each package discovered, check up on installer-paths to adjust its root path.

Each path can contain `{$name}`, `{$vendor}` and `{$type}`. Vendor can be taken from the package name, name is either taken from `"installer-name"` from the `"extra"` section of te packages composer.json, or the package name. Type is the `"type"` from the composer.json or `"library"`...

Each path points to an array of what to install there, which can be:
1. A package name, ie. `"drupal/core"`
2. A vendor: `"vendor:my_organization"`
3. A package type: `"type:drupal-module"`

A bit fiddly, but not too bad...

This is step one in getting Drupal 8 working. Of course a solution that just hooks into ede-php-autoload and adjusts things for Drupal is also a possibility, but supporting composer/installers gives support for many more frameworks.

**Comment by xendk** [comment]
I've started constructing a "project" in the test folder. I love TDD. It'll probably be slow progress over the Christmas period, but better slow progress than no progress.

**Comment by stevenremot** [comment]
I totally agree with your approach, supporting composer/installers is definitely a good thing! What about creating the PR early, so that I can give you my feedback on the code in "real time"?

**Comment by xendk** [comment]
Sure. I'll create the PR when I've stubbed out the project.

**Comment by xendk** [comment]
Bad news: composer/installers basically screws us over. I'd overlooked the fact that the `installer-paths` key in the `composer.json` file I was looking at is just a local override. Each framework installer in composer/installers supplies their own default paths for all their types.

So if one is using the drupal installer, and are happy with the default path that it supplies, there's no hint in the `composer.json` nor `composer.lock` where it was installed. To add insult to injury, installers might munge the name of the package in unpredictable ways.

Pondering what to do about that, as being unable to find the file of classes kills off all the fun of semantic-php...

1. One could implement the logic I outlined in the issue, and in addition to that implement the logic of all the composer installers. The downside is that it's much work, and ede-php-autoload would have to track composer/installers and implement changes in a timely manner.
2. One could fall back to the tags approach and scan all files and make an index of classes to files. But it's slow and annoying.
3. Petition composer or composer/installers to note the packages install dirs in the lock file. Probably the simplest from our point if view, but who knows if they'll like the idea.
4. Use a php script to extract the class/namespace/psr4 maps from vendor/autoload_*.php. Requires a working php command locally, and is sensitive towards changes in how composer constructs its autoloaders.

What do you think?

**Comment by stevenremot** [comment]
Yo, for now I'm trying to understand what composer/installers is doing exactly!

If there is too much work to do to implement installers, I think it will he better to just implement some hooks in ede-php-autoload and implement all the composer/installers logic u n another package to keep this one focused. 

But I have another suggestion: what about generating a classmap with composer and declare this classmap in your project? ede-php-autoload already has support for classmap so this can be a good first step.

Regarding your suggestion on tags, you will be better served with another semanticdb backend than ede-php-autoload if you go this way.

**Comment by stevenremot** [comment]
I had a better look at it. What about forgetting composer/installers and implementing drupal support directly? I think supporting composer/installers doesn't really worth it, as it would require doing a lot of of specific code for each installer in the end.

If you tell me what you need exactly, we can see which hooks we can put in the ede-php-autoload code to better integrate new project types, and especially drupal.

**Comment by xendk** [comment]
Well, classmap could work, but, as I understand it, it would require running composer after adding a new class. I'd rather prefer something less manual.

Thing is, doing Drupal (and other framework) support completly outside of ede-php-autoload is either going to be half-hearted or duplicate much of ede-php-autoload. A naive implementation would have to guess whether `drupal/core` is installed in `core` or `web/core` and fail miserably when someone decides on `docroot/core`. Trying harder would mean reading the `composer.json` and checking the `installer-paths`, which would be more elegantly done with some support from `ede-php-autoload`.

The thought about using the old filescan method came from pondering how IDEs like PHPStorm might do it. I suspect that might actually be what they're doing, but with the advantage of smart use of file notifications and background scanning. Not easily duplicated.

It still seems to me that the simplest solution would be for composer to note the installed path of packages in `composer.lock`. Whether the composer team would agree is another question, possibly they'd object to ede-php-autoload reading the lock file in the first place.

What I was thinking might be realistically workable (save help from composer), was for ede-php-autoload to  understand `composer/installers` enough to:

1. Notice the `type` property in the lock file and
2. Either using the path specified in `intaller-paths`, if one is given,
3. Or map up types to directories using a customize variable so end users can add the common paths themselves.
4. Let a hook fiddle with it for the odd case where a simple mapping wont do.

I think 1 to 3 will cover 99% of all users needs, and the last hook gives people something to work with.

Given the right hooks in ede-php-autoload, it could be implemented in a ede-php-composer-installers package. The minimum required would be for ede-php-autoload to hand over the package name and the type from the lock file, and get back a base directory to use instead of `vendor/<vendor>/<package>`. I could work with that at least.

My lisp-fu is weak, but isn't `ede-php-autoload-composer--get-third-party-dir` basically the place to hook in? I figure I could just redefine it with my own function to make it do what I want, though a more proper hooking mechanism would be preferable. The only problem with that is figuring out whether `composer/installers` are installed by the `composer.json`, as nothing special should be done in the case where it isn't.

Oh, and bear in mind that there's seems to exist other custom installers works much like `composer/installers` but are for a specific framework (`typo3/cms-composer-installers`).

So a `ede-php-autoload-composer--get-third-party-dir` hook basically needs to have the list of packages form `composer.json` (too see if it's installed), the `extras` section from same (it's where customizations are specified), the package name and the `composer.lock` section for the package. The first two could be handed to some sort of init function, but the init function would need to have some place to store state in the project for the hook.

Can you think up a nice API for that?

**Comment by stevenremot** [comment]
I was actually thinking about a more general hook to be able to parse any arbitrary configuration from the composer file and add the necesary autoloads. It would let you define a new composer "visitor" like this:

```emacs-lisp
(defun my-custom-composer-visitor (composer-data composer-lock base-directory autoloads)
  ;; Add your own autoloads in the autoloads var, or clone it and enrich it
  autoloads)

(ede-php-autoload-composer-define-visitor #'my-custom-composer-visitor)
``` 

The ede-php-autoload internal code will use it too. For example, `ede-php-autoload--merge-composer-autoloads` would be rewritten like this:

```emacs-lisp
(defun ede-php-autoload--merge-composer-autoloads (composer-data composer-lock base-dir autoloads)
  "Load the autoload information in COMPOSER-DATA and merge it with AUTOLOADS.

COMPOSER-DATA is the parsed composer.json file.
BASE-DIR is the prefix dir to add to each autoload path."
  (let ((composer-autoloads (append (cdr (assoc 'autoload composer-data))
                                    (cdr (assoc 'autoload-dev composer-data))))
        key spec base-spec)
    (dolist (autoload-part composer-autoloads)
      (when (member (car autoload-part) '(psr-0 psr-4))
        (setq key (intern (concat ":" (symbol-name (car autoload-part))))

              spec (mapcar #'(lambda (pair)
                               (ede-php-autoload--format-composer-pair pair base-dir (car autoload-part)))
                           (cdr autoload-part))

              base-spec (plist-get autoloads key))

        (setq autoloads (plist-put autoloads key (append base-spec spec)))))
    autoloads))

(ede-php-autoload-composer-define-visitor #'ede-php-autoload--merge-composer-autoloads)
``` 

(Nothing changes except the argument list)

`ede-php-autoload--merge-lock-file-data` would be rewritten to use the same system.

`ede-php-autoload--append-composer-data` would be rewritten to be a big reduce on all visitors.

This way, it will be possible to create a new emacs package that defines its own visitor to scan the composer.json and add its own Psr0 / Psr4 / whatever (you can create your own class loaders with `ede-php-autoload-class-loader-define-factory`). You can create a visitor that searches the composer entries for the "type" property and add the autoload entries corresponding to the project type.

The only thing I'm not sure is the argument list. I think sending a alist may be better than sending 4 arguments, because this list of arguments could grow with time and become difficult to manage...

Regarding your suggestion on having the info in the composer.lock file, I think this is worth discussing with the composer developers. If they don't agree with us scanning the composer.lock file, they may give us a better solution.

Anyway, do you think the API is good enough to allow other packages extending `ede-php-autoload`? If so, I will implement it even if we're not sure you will have to use it. It will only make the code better :-)

**Comment by xendk** [comment]
Well yeah, I figure that the API makes it possible to implement the ede-php-composer-installers previously outlined, though it's a bit difficult to follow for a lisp-beginner like me.

Two things:

1. Order of visitors? I'll need to get in after the one that loads the lock file, and other visitors might have other requirements.
2. You talk about adding class loaders, but I just need to adjust the paths of the ones that `ede-php-autoload--merge-lock-file-data` already added. Will I have to loop through autoloads to find the ones I need to fix up?

**Comment by stevenremot** [comment]
To answer your questions:

1. All visitors will be provided with composer.json and composer.lock content, already parsed in a lisp structure
2. Mmmmmh, it depends, I don't really understand the details in composer/installers. Will you have to "fix" wrong autoloads, or to create brand new ones? If you have to create new ones, it will be easy. Your visitor will receive the current autoloads set by previous visitors. You just have to append the new ones, and to return the new list of autoloads in your visitor.

**Comment by xendk** [comment]
Fix wrong autoloads. All composer/installers really do is change the path of the package directory from `vendor/<vendor>/<package>` to, say `web/modules/contrib/<package-name>`. Unless ede-php-autoload checks the directory for existence, it'll just think the package is in the wrong place.

So unless the autoloads are indexed in some way, I'll have to loop through the composer.lock data and then search through autoloads for the found projects. Sounds a bit ineffective to me.

**Comment by stevenremot** [comment]
So actually the autoloads created by composer/installers can be found in the composer.lock? I didn't understand that. But you are right, this can be ineffective for big composer.lock files...

If I understand correctly, in the case of Drupal, the third-party packages are put out of vendor, and Drupal itself takes care of autoloading without the help of composer, right? If that's the case, what about forgetting composer/installers totally, and implement the Drupal autoload logic? I don't know Drupal enough to know if it's a sane solution or not.

**Comment by xendk** [comment]
That's correct. I believe there's no difference to the composer.lock, with or without composer/installers.

Only partly. Drupal uses the composer autoloader for the core classes, but these are moved to, for instancew `core/lib` by composer/installers. How composer/installers does this, I don't know, but I can see Drupal make a showing in `vendor/composer/autoload*`

How it loads classes of modules (which might not be installed by composer, might even not have a composer.json file) I haven't figured out yet. But yeah, I'll have to figure out an autoloader for that.

But I see where you're going. If I implement the autoloading for modules, it could also handle Drupal core. But then I'd need to suppress the Drupal autoloads created by ede-php-autoload-composer, as they'd still point down at `vendor/`. Or fix them up. But I'll still need the composer autoloads for non-Drupal libraries.

**Comment by stevenremot** [comment]
I have an idea in order to let you suppress wrong autoloads created by ede-php-autoload. It should not be abused, but what do you think about being able to give priority to visitors, so that you could define a visitor that would be run after the standard composer stuffs ? You will be able to remove the wrong autoload composer created, and to add your drupal autoloads.

It would be a definition like this:

```emacs-lisp
(ede-php-autoload-composer-define-visitor
 #'add-drupal-autoloads
 :after)
```

(I'm not really sure how to name priorities)

**Comment by xendk** [comment]
I've got a bit of difficulty picturing things in my head by now, but I guess yeah? Can't I just adjust the path of the "wrong" autoloads? They're still valid, it's just the path.

For a first stab, I can just do a visitor that adjusts packages of certain known types to the path that suits me. If I can get that working, using the installer-paths from the composer file is the next step.

**Comment by stevenremot** [comment]
Yes, you can totally just add your new paths and keep the old, wrong ones! It will just be a path with no directory, this shouldn't raise any error. Then I come back to you when I refactored my code to add the visitor stuffs. I don't handle priorities for now.

**Comment by stevenremot** [comment]
Hey @xendk !

I come back to you now that I implemented the visitor system. I put it in a PR so that you can check it easily: https://github.com/stevenremot/ede-php-autoload/pull/16

**Comment by xendk** [comment]
Looks good to me. A visitor gets all the data I need as far as I can tell, and can munge, add to or delete from the autoloads as needed.

I'll work on a package that uses it, but I might need to consult my local elisp guru. Luckily he's back from vacation tomorrow. :-D

**Comment by stevenremot** [comment]
You have a personal elisp guru? That sounds something really cool to have at hand :-D

I'm merging the PR then

**Comment by xendk** [comment]
Got a version running passing tests: https://github.com/xendk/ede-autoload-installers/

Comments welcome.

It doesn't handle typed packages without a specific path in the composer.json yet (planning on a customize setting to supply default paths for types), or packages supplying their own name. And I haven't tried it out on a live project yet.

I would like if `ede-php-autoload--merge-composer-autoloads` would sort out duplicates. It feels iffy to have duplicates and rely on the missing file.

**Comment by stevenremot** [comment]
Good job, that's much shorter than I thought it would be!

I see that you are using a lot of functions that are private to `ede-php-autoload` (those which have two dashes in the name). I will try to make a public API that fulfill your needs. Maybe just make these functions public, I'll see.

And I can definitely avoid duplicates in the merge function, that would be much better!

**Comment by xendk** [comment]
Ah, that's what the double dashes mean. :-)

Yes, I'd rather like not to have to reimplement those functions.

**Comment by xendk** [comment]
Oh, and I was wondering if you have some ede/ede-php-autoload tricks? It's a bit of an issue for me that I don't know what ede can actually do.

After determining that `(find-file (ede-php-autoload-find-class-def-file (ede-current-project) "Drupal\\Core\\Database\\Schema"))` works excellently, I pondered making an ido based function to open a file from the class name, but wondered if that's not something ede can already do.

**Comment by stevenremot** [comment]
My understanding of EDE is actually somewhat limited too, but as far as I know, it's more about managing files, compilation targets, etc... No semantical stuffs. But semantic should be able to do this, that's actually why I implemented a semanticdb backend in `ede-php-autoload`. When semantic is enabled, you may be able to find your class with `C-c , J`. I didn't try, so I can't be sure...

**Comment by stevenremot** [comment]
I started working on your requests : https://github.com/stevenremot/ede-php-autoload/pull/17

The public API is ready. I just renamed the functions you are using to allow their use outside of the package.

I prepared ert-runner to be able to code the removing of the duplicates in the autoloads safely.

**Comment by xendk** [comment]
I think that `C-c , J` uses the data that semantic has loaded? And it is a bit odd. But I got thinking that as ede-php-autoload doesn't actually know about classes until asked, it's probably best implemented as a specific function.

Re #17: I think I'm being stupid. If i can `(append)` `require` and `require-dev` for the composer data, I can do the same for `packages`. But why are you using `(vconcat)` in `ede-php-autoload-composer--get-third-party-data` and `(append)` in `ede-php-autoload-composer-create-autoloads-from-data`? 

So I think `ede-php-autoload-composer--get-third-party-data` should stay private.

I should probably also make some ert tests to unit test some of the internal functions.

**Comment by xendk** [comment]
I've eliminated my need for `ede-php-autoload-composer--get-third-party-data`, so the only function I need public is `ede-php-autoload--merge-composer-autoloads`.

**Comment by xendk** [comment]
Side note: I've figured out how Drupal adds in autoloaders for modules. The solution seems to scan some pre-defined directories for folders, determine which is probably a module (checking for `folder/src`might be acceptable, checking for `folder/folder.info.yml` should be pretty sure), and adding `folder/src` as PSR4 for the `Drupal\folder` namespace (perhaps another namespace too). I'll probably make an `ede-autoload-drupal` package for that, but I'd rather that ede-autoload-installers visitor was run first, as I can inspect autoloads to determine where the Drupal root is, without having to reimplement the path adjusting code of ede-autoload-installers, but using `ede-php-autoload-composer-define-visitor` and usual requiring would make the Drupal visitor run first. So I need a way to ensure the proper order.

BTW, what are your thoughs about my naming of the package? I'm getting second thoughts, should I have called it `ede-php-autoload-installers` to make the dependency more explicit, and to make the packages show up next to each other when I get around to releasing it on melpa.

**Comment by stevenremot** [comment]
> I think that C-c , J uses the data that semantic has loaded? And it is a bit odd. But I got thinking that as ede-php-autoload doesn't actually know about classes until asked, it's probably best implemented as a specific function.

Maybe a tag system would be more interesting then ? The idea behind `ede-php-autoload` was that I wanted to handle external references without having to rely on an indexing step. This means there are some operations like listing all classes in a project that are not possible, and will never possible with this package.

> Re #17: I think I'm being stupid. If i can (append) require and require-dev for the composer data, I can do the same for packages. But why are you using (vconcat) in ede-php-autoload-composer--get-third-party-data and (append) in ede-php-autoload-composer-create-autoloads-from-data?

This part is a contribution so I can't be sure, but I may have been done like because `append` mutates its last argument. I agree it's not consistent with the rest of the code.

> I've eliminated my need for ede-php-autoload-composer--get-third-party-data, so the only function I need public is ede-php-autoload--merge-composer-autoloads.

OK, I will hide it again then!

> Side note: I've figured out how Drupal adds in autoloaders for modules...

Mmmmh, so basically your drupal package will depend on your installers package to ensure the base paths are setup... If this is an extension to `ede-autoload-installers`, wouldn't it be safer to make `ede-autoload-installers` extensible? That will make the dependency clearer... The only drawback I see is that we have a really funny chain of hooks and dependencies this way, but the problem itself has a funny chain of dependencies...

> BTW, what are your thoughs about my naming of the package? I'm getting second thoughts, should I have called it ede-php-autoload-installers to make the dependency more explicit, and to make the packages show up next to each other when I get around to releasing it on melpa.

Ah, you're right, it maybe better! I will add an entry in my README once you feel your package is ready for the release anyway.

**Comment by xendk** [comment]
> Maybe a tag system would be more interesting then ? 

GNU Global was already causing Emacs to hang while re-indexing (ggtags tries to incrementally update the index on file save), and it doesn't understand interfaces. And like you, I'd rather like to avoid the indexing.

You're right, with autoloads we can't list classes like that, but I was thinking of a command to complete through namespaces, and then switch to file scanning after that. Actually, a simple but workable solution would be a function that first completes an namespace from the autoload list and then just starts `find-file` from the selected namespaces directory.

Another possibility I just thought up, which might actually fulfill the practical need better, is getting a list of all classes mentioned in the current file, do an ivy/helm/whatever completion on that and load the selected class. Assuming that semantic-php has figured out the file properly.

>  If this is an extension to ede-autoload-installers, wouldn't it be safer to make ede-autoload-installers extensible? 

Well, it isn't _really_. The `ede-php-autoload-drupal` will need to know where drupal/core is installed, to figure out the paths to module directories, but that's it. Feels wrong to bolt it onto `ede-php-autoload-installers`.

> Ah, you're right, it maybe better!

OK, I'll rename it. Really it should be `ede-php-autoload-composer-installers` but it's ab bit ridiculous and would make it the package with the longest name, with `dirtrack-buffer-name-track-mode` a close second. Though, I think I'll go with that as it does spell out it's two dependencies out loud, and I'll just have to live with the long symbol names. :-)

Don't know when I'll get it fixed, I'm on vacation this week, and family takes precedence.

Anyway, guess we can close this issue. I still need to support package name munging to call the package complete, but it basically works. And after `ede-php-autoload-drupal`, I'll finally be able to get back to try and improve on my semantic-php fork (sadly https://github.com/jorissteyn/semantic-php seems a bit dead).

**Comment by stevenremot** [comment]
> You're right, with autoloads we can't list classes like that, but I was thinking of a command to complete through namespaces, and then switch to file scanning after that. Actually, a simple but workable solution would be a function that first completes an namespace from the autoload list and then just starts find-file from the selected namespaces directory.

Hey, actually this is already implemented in ede-php-autoload! The method is called `ede-php-autoload-complete-type-name`, but it's more interesting to use it from the semanticdb backend when possible, using `semanticdb-find-tags-by-name-method`. That way you can change the completion backend at will.

> Well, it isn't really. The ede-php-autoload-drupal will need to know where drupal/core is installed, to figure out the paths to module directories, but that's it. Feels wrong to bolt it onto ede-php-autoload-installers.

What about exposing the installer path resolving system as an API then?

When I was talking about a priority system, I though about something simple, like "run before standard stuffs" and "run after standard stuffs". But I think priority between different external visitors will add complexity to the code, and also to the ecosystem: if you have to think about which visitor will be executed before and after the visitor you are writing, it's going to be hard to maintain. That's why I would like to find an alternate solution, even if I have to develop a bit more helper functions to make it work.

> OK, I'll rename it. Really it should be ede-php-autoload-composer-installers but it's ab bit ridiculous and would make it the package with the longest name, with dirtrack-buffer-name-track-mode a close second. Though, I think I'll go with that as it does spell out it's two dependencies out loud, and I'll just have to live with the long symbol names. :-)

Yeah, that's hard to find a proper name for packages... Did you try [nameless](https://github.com/Malabarba/Nameless)? At least it helps removing all the noise with package names when you develop.

> Anyway, guess we can close this issue.

OK, I will merge my PR so that you can use it then! I will reopen a PR for the improvements of `ede-php-autoload-composer-merge-autoloads`.

> sadly https://github.com/jorissteyn/semantic-php seems a bit dead

Totally. Even if I don't write PHP code anymore, I'm happy you are trying to improve PHP editing experience in Emacs. It's really hard to convince your colleagues you are productive when it's missing so much features comparing to IDEs :-D

**Comment by xendk** [comment]
> `semanticdb-find-tags-by-name-method`

Ah, just need to figure out how to use that then.

> What about exposing the installer path resolving system as an API then?

Well, the Drupal extension just need to know where Drupal core is installed, it's indifferent to how it got there. And while Drupal seems to have been married to composer/installers while I wasn't paying attention, other frameworks might not be. TYPO3 seems to have its own custom installer, but there's notthing stopping another framework from using standard vendor for installing and then adding some custom plugin paths to the autoloader.

The way I see it, there's two kind of visitors, "early" and "late". The "early" visitor is composer/installers and probably the TYPO3 installer. They should come in early as they basically changes the layout of the project. And there's likely just one "early" visitor per project, The "late" visitor is the Drupal visitor, possibly others which adds stuff that's actually outside composer, but added by the framework to the autoloader in their bootstrap process. They're likely to trigger on a specific package.

So I think being able to both prepend and append to the list is sufficient. I think that in most cases where visitors is implicitly depending on each other, you'll have one developer being aware of it.

> Did you try nameless? 

Considered it, but I'll rather just live with the long names. 

> It's really hard to convince your colleagues you are productive when it's missing so much features comparing to IDEs :-D

And being ridiculed. Thankfully, while we're still a small shop (~20 people), we're two Emacs users, and the other is the maintainer of drupal-mode, so..

**Comment by xendk** [comment]
Well, ede-php-autoload-compsoer-installers is pretty much ready, and ede-php-autoload-drupal is pretty close too.

How about just adding an argument to `ede-php-autoload-composer-define-visitor` to append rather than prepend the visitor? I think the need to control the order of different visitors is pretty limited outside of my case and it's a quick solution.

All I need now is to get it into melpa.

**Comment by stevenremot** [comment]
> Well, ede-php-autoload-compsoer-installers is pretty much ready, and ede-php-autoload-drupal is pretty close too.

Good job!

> How about just adding an argument to ede-php-autoload-composer-define-visitor to append rather than prepend the visitor? I think the need to control the order of different visitors is pretty limited outside of my case and it's a quick solution.

I think I will come with a proper priority system. I'm afraid just appending or prepending to the list will produce problems because we can't really now in which order files will be loaded. I will have a `:early` and `:late` argument as you asked, but they will be properly separated groups. I will try to implement it in the week!

**Comment by xendk** [comment]
`:early` and `:late` works for me! It could default to `:late` as I think that's the most useful for the general case.

**Comment by stevenremot** [comment]
I wanted to default to neither `:early` or `:late`. Something like `:normal`, in between. This way it will cover more use cases, and hopefully I won't have to add complexity to this system later.

**Comment by xendk** [comment]
Aha, I was actually thinking of having a middle ground, but thought you'd dislike another option. :-)

Don't think you'll have to add any more later. I think my usecase is as messy as they come.

**Comment by stevenremot** [comment]
Here is my implementation : https://github.com/stevenremot/ede-php-autoload/pull/18 . If it's okay to you I will merge it.

**Comment by xendk** [comment]
:+1: Looks good to me.

**Comment by stevenremot** [comment]
Merged

**Comment by xendk** [comment]
Excellent, I'm looking into the ceremony involved in getting ede-php-autoload-composer-installers into melpa right now. Hope I survive the code review.

By the way, did you get round to implementing de-duplication in `ede-php-autoload-composer-merge-autoloads` or should I create a new issue? It's not critical, but would be a nice cleanup.

Anyway, we can really close this issue now. :-)

**Comment by stevenremot** [comment]
> Excellent, I'm looking into the ceremony involved in getting ede-php-autoload-composer-installers into melpa right now. Hope I survive the code review.

Aha, I think you'll be okay :-)

> By the way, did you get round to implementing de-duplication in ede-php-autoload-composer-merge-autoloads or should I create a new issue? It's not critical, but would be a nice cleanup.

I started writing tests for it, but it's not implemented yet. I will open an issue for it.

> Anyway, we can really close this issue now. :-)

I will click the "Close and comment" button then!
